### PR TITLE
PagePrefetcher: compute PageIds only when needed

### PR DESCRIPTION
**Pull request by gab8i**

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/56?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#56** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/56?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#55** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/55?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: nomt/src/cursor.rs [hunk 1]
```diff
@@ -315,16 +308,19 @@ impl PagePrefetcher {
...
-            let n = cmp::min(PREFETCH_N, self.rev_page_ids.len());
-            for _ in 0..n {
-                // Unwrap: pop must succeed since the n is clamped to the length of the vector.
-                let page_id = self.rev_page_ids.pop().unwrap();
+            for i in 0..PREFETCH_N {
+                let page_id = match self.page_ids_iter.next() {
+                    Some(page) => page,
+                    // if page_ids_iter returns none and we are looking for the
+                    // first PageId of the new batch of pages then there are no pages left to prefetch
+                    None if i == 0 => return None,
```
Comment (rphmeier) [inline nomt/src/cursor.rs:316]: is this true for `PREFETCH_N != 7`?

File: nomt/src/cursor.rs [hunk 2]
```diff
@@ -298,15 +298,8 @@ struct PagePrefetcher {
...
-        let rev_page_ids = {
-            // Chop the destination key path into the corresponding page IDs and reverse them so
-            // that we can `pop`.
-            let mut x = PageIdsIterator::new(dest).collect::<Vec<_>>();
-            x.reverse();
-            x
-        };
         Self {
-            rev_page_ids,
+            page_ids_iter: PageIdsIterator::new(dest),
```
Comment (pepyakin) [inline nomt/src/cursor.rs:302]: Could we please `reserve(DEPTH)` while we are at it? 🥺👉👈

Comment (rphmeier) [reply nomt/src/cursor.rs:302]: do you mean `page_promises: Vec::with_capacity(PREFETCH_N)`?

Comment (pepyakin) [reply nomt/src/cursor.rs:302]: Yup!

**Review by rphmeier — APPROVED** [review]

**Pull request merged**